dali/daliadmin/daliadmin.cpp

```diff
@@ -2794,114 +2798,183 @@ static void wuidCompress(const char *match, const char *type, bool compress)
+                    IpAddress ip(host);
+                    StringBuffer ipStr;
+                    ip.getIpText(ipStr);
+                    const auto &it = machineMap.find(ipStr.str());
+                    if (it == machineMap.end()) // not found, remove it.
+                    {
+                        if (verbose)
+                            PROGLOG("Removing stale for %s at location %s", name, host);
+                        locationsToRemove.push_back(&location);
+                        ++removedLocations;
+                    }
                 }
             }
+            if (fix)
+            {
+                for (auto &location: locationsToRemove)
+                    gd.removeTree(location);
+            }
         }
-        ++index;
     }
-    PROGLOG("%d out of %d workunit files not associated with any workunit. Took %d ms", indexToDelete.ordinality(), index, ts.elapsed());
-    ts.reset(0);
+    PROGLOG("%u out of %u generated dll entries not associated with any workunit, and %u stale locations found. Took %u ms", (unsigned)generateDllsToDelete.size(), totalCount, removedLocations, ts.elapsed());
 
+    ts.reset(0);
     IArrayOf<IDllEntry> removedEntries;
     unsigned numDeleted = 0;
-    ForEachItemInRev(d, indexToDelete)
-    {
-        const char *uid = uidsToDelete.item(d);
-        unsigned index = indexToDelete.item(d);
-        StringBuffer path("GeneratedDll[");
-        path.append(index).append("]");
-        IPropertyTree *gd = root->queryPropTree(path.str());
-        if (NULL == gd)
-            throwUnexpected();
-        const char *uidQuery = gd->queryProp("@uid");
-        if (0 != strcmp(uid, uidQuery))
-            throw MakeStringException(0, "Expecting uid=%s @ GeneratedDll[%d], but found uid=%s", uid, index, uidQuery);
+    for (auto &toDelete: generateDllsToDelete)
+    {
         if (verbose)
-            PROGLOG("Removing: %s, uid=%s", path.str(), uid);
+            PROGLOG("Removing: %s", toDelete->queryProp("@name"));
         if (fix)
         {
-            Owned<IDllEntry> entry = queryDllServer().createEntry(root, gd);
-            entry->remove(false, false); // NB: This will remove child 'gd' element from root (GeneratedDlls)
+            Owned<IDllEntry> entry = queryDllServer().createEntry(root, toDelete);
+            entry->remove(false, false); // NB: this is equivalent to root->removeTree(&toDelete);
             if (deleteFiles) // delay until after meta info removed and /GeneratedDlls unlocked
                 removedEntries.append(*entry.getClear());
         }
         ++numDeleted;
     }
     if (fix)
     {
-        conn->commit();
-        PROGLOG("Removed %d unassociated file entries. Took %d ms", numDeleted, ts.elapsed());
-        ts.reset(0);
+        generatedDllsConn->commit();
+        PROGLOG("Removed %d unassociated file entries, and %u stale locations found. Took %u ms", numDeleted, removedLocations, ts.elapsed());
```

Comment:
will change slightly, but leave in the repetition of the information, since the initial report can be quick/early and the final tracing is useful summary.